src/library_loader.h

```diff
@@ -0,0 +1,115 @@
+/*
+ * This file is part of OpenTTD.
+ * OpenTTD is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 2.
+ * OpenTTD is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file library_loader.h Functions/types related to loading libraries dynamically. */
+
+#ifndef LIBRARY_LOADER_H
+#define LIBRARY_LOADER_H
+
+#include <string>
+
+class LibraryLoader {
+public:
+	/**
+	 * A function loaded from a library.
+	 *
+	 * Will automatically cast to the correct function pointer type on retrieval.
+	 */
+	class Function {
+	public:
+		explicit Function(void *p) : p(p) {}
+
+		template <typename T, typename = std::enable_if_t<std::is_function_v<T>>>
+		operator T *() const
+		{
+			return reinterpret_cast<T *>(this->p);
+		}
+
+	private:
+		void *p;
+	};
+
+	/**
+	 * Load a library with the given filename.
+	 */
+	explicit LibraryLoader(const std::string &filename)
+	{
+		this->handle = this->OpenLibrary(filename);
+	}
+
+	/**
+	 * Close the library.
+	 */
+	~LibraryLoader()
+	{
+		if (this->handle != nullptr) {
+			this->CloseLibrary();
+		}
```

Comment:
Not quite sure about this destructor.

On one hand having the separate `CloseLibrary` function means this `nullptr` check can be left out of all the implementers, but it also triggers the question: shouldn't `CloseLibrary` set the handle to `nullptr`, just in case someone changes the internal logic of this at some later point.

On the other hand, the different OS specific implementations can also just implement the destructor and essentially rename `CloseLibrary` to `~LibraryLoader`. Though that'll make the constructor feel weird, but the `OpenLibrary` implementations could also become the constructor in the implementations.

Reply:
Tomato tomato. Nothing wrong with the current choice, so going to leave it as it is :) I don't feel like retesting everything for just this, sorry :)

Reply:
unique_ptr with custom destructor?